File: python/people4_invoice_cli.py
"""
People4 Invoice API CLI

Retrieves a JWT from the People4 Authentication API, then sends an invoice
payload (JSON) to the People4 Invoice API and prints the UBL/PEPPOL result.

Usage:
    python python/people4_invoice_cli.py
"""

from __future__ import annotations

import json
import ssl
import sys
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Optional


class People4HttpClient:
    def __init__(self, disable_ssl_verification: bool = False) -> None:
        self._disable_ssl_verification = disable_ssl_verification

    # send an invoice and receive ubl, peppol and people4Id in response
    def send_invoice(self, url: str, payload: str, jwt: str) -> dict:
        response = self._http_request(url, "POST", payload, jwt)
        if not response:
            raise RuntimeError("Response failed to retrieve JWT token from People4 Authentication API.")
        return json.loads(response)

    def retrieve_token(self, url: str, payload: "RetrieveTokenPayload") -> str:
        response = self._http_request(url, "PUT", json.dumps(payload.to_dict()), payload.api_key)
        if not response:
            raise RuntimeError("Response failed to retrieve JWT token from People4 Authentication API.")
        token = json.loads(response).get("token")
        if not token:
            raise RuntimeError("JWT token not present in response.")
        return token

    def _http_request(self, url: str, method: str, payload: str, bearer: str) -> str:
        body = payload.encode("utf-8") if payload else None
        request = urllib.request.Request(
            url=url,
            data=body,
            method=method,
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
                "Content-Length": str(len(body) if body else 0),
                "Authorization": f"Bearer {bearer}",  # use for jwt and api key
            },
        )
        context = None
        if self._disable_ssl_verification:
            # curl -k insecure on DEVELOPMENT, not recommended for production
            context = ssl.create_default_context()
            context.check_hostname = False
            context.verify_mode = ssl.CERT_NONE
        try:
            with urllib.request.urlopen(request, timeout=15, context=context) as response:
                return response.read().decode("utf-8")
        except urllib.error.HTTPError as exc:
            error_body = exc.read().decode("utf-8")
            raise RuntimeError(f"HTTP request failed with status code {exc.code}: {error_body}") from exc
        except urllib.error.URLError as exc:
            raise RuntimeError(f"HTTP request failed: {exc.reason}") from exc


@dataclass(frozen=True)
class RetrieveTokenPayload:
    api_key: str
    domain: str
    legal_name: str
    vat_id: str

    def to_dict(self) -> dict:
        return {
            "domain": self.domain,
            "legalname": self.legal_name,
            "vatId": self.vat_id,
        }


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------


class People4InvoiceCli:
    def __init__(self) -> None:
        self._jwt: Optional[str] = None

    # retrieve JWT token from Authentication API
    def get_token(self) -> str:
        http_cli = People4HttpClient(disable_ssl_verification=True)  # only for testing, not recommended for production
        payload = RetrieveTokenPayload(
            api_key="1c8f86e857cb65b20b26c481b0e2d2d1bf0b93a93aa7ccc35704c37ee63c597b",
            domain="acme-corporation.eu",
            legal_name="ACME Corp",
            vat_id="NL8200.98.395.B.01",
        )
        self._jwt = http_cli.retrieve_token(
            url="https://app.people4.eu/auth-api/token",
            payload=payload,
        )
        return self._jwt

    # send invoice payload to Invoice API POST method
    def create_ubl_invoice(self, invoice_json_raw: str) -> dict:
        http_cli = People4HttpClient(disable_ssl_verification=True)  # only for testing, not recommended for production
        return http_cli.send_invoice(
            url="https://app.people4.eu/invoice-api/invoice/v1",
            payload=invoice_json_raw,
            jwt=self._jwt,
        )


def main() -> int:
    easy_compliance_cli = People4InvoiceCli()
    try:
        print("Trying to retrieve JWT token from Authentication API")
        easy_compliance_cli.get_token()
        # print(f"OK: JWT token retrieved: '{jwt}'")

        print("Read json invoice minimal test file and send to People4 Invoice API")
        invoice_file = Path(__file__).parent.parent / "inputs" / "invoice-v1-minimal-test.json"
        invoice_json_raw = invoice_file.read_text(encoding="utf-8")
        print("Trying to send invoice payload to Invoice API to generate UBL and PEPPOL invoice")
        invoice_created = easy_compliance_cli.create_ubl_invoice(invoice_json_raw)
        # print(f"DEBUG: Invoice created response: {json.dumps(invoice_created)}")

        if "people4Id" not in invoice_created:
            raise RuntimeError("People4 ID not present in response.")
        if "ubl" not in invoice_created:
            raise RuntimeError("UBL not present in response.")
        if "peppol" not in invoice_created:
            raise RuntimeError("PEPPOL not present in response.")

        messages = invoice_created.get("messages")
        if messages:
            message = None
            for message in messages:
                print(f"Error Message: {message}")
            raise RuntimeError(f"Error on generate UBL or PEPPOL message: {message}")

        if invoice_created.get("people4Id"):
            print(f"OK: Invoice created with People4 ID: {invoice_created['people4Id']}")
        if invoice_created.get("ubl"):
            print(f"OK: UBL invoice created with length: {len(invoice_created['ubl'])}")
        if invoice_created.get("peppol"):
            print(f"OK: PEPPOL invoice created with length: {len(invoice_created['peppol'])}")
        return 0
    except Exception as exc:  # noqa: BLE001
        print(f"Fatal: {exc}", file=sys.stderr)
        print(f"Error: {exc}")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: php/People4InvoiceCli.php
<?php

declare(strict_types=1);

class People4HttpClient
{
    public function __construct(
        private readonly bool $disableSslVerification = false,
    ) {}

    // send an invoice and received ubl, peppol and people4Id in response
    public function sendInvoice(string $url, string $payload, string $jwt): array
    {
        $response = $this->httpRequest(
            $url,
            'POST',
            $payload,
            $jwt,
        );
        if (!$response || !is_string($response)) {
            throw new \RuntimeException("Response failed to retrieve JWT token from People4 Authentication API.");
        }
        $invoiceCreated = json_decode($response, true, 512, JSON_THROW_ON_ERROR);
        return $invoiceCreated;
    }

    public function retrieveToken(string $url, RetrieveTokenPayload $payload): string
    {
        $response = $this->httpRequest(
            $url,
            'PUT',
            json_encode($payload->toArray(), JSON_THROW_ON_ERROR),
            $payload->getApiKey(),
        );
        if (!$response || !is_string($response)) {
            throw new \RuntimeException("Response failed to retrieve JWT token from People4 Authentication API.");
        }
        $jwt = json_decode($response, true, 512, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE)['token'] ?? throw new \RuntimeException("JWT token not present in response.");
        return $jwt;
    }

    private function httpRequest(string $url, string $method, string $payload, string $bearer): mixed
    {
        $ch = curl_init();

        curl_setopt_array($ch, [
            CURLOPT_URL            => $url,
            // CURLOPT_POSTFIELDS     => $json,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 15,
            CURLOPT_HTTPHEADER     => [
                'Content-Type: application/json',
                'Accept: application/json',
                'Content-Length: ' . strlen($payload),
                'Authorization: Bearer ' . $bearer, // use for jwt and api key
            ],
            CURLOPT_HEADER         => false, // Collect response headers
            CURLOPT_SSL_VERIFYPEER => !$this->disableSslVerification, // curl -k insecure on DEVELOPMENT, not recommended for production
            CURLOPT_SSL_VERIFYHOST => !$this->disableSslVerification, // curl -k insecure on DEVELOPMENT, not recommended for production
        ]);
        if ($payload) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, $payload);
        }
        if ($method === 'PUT') {
            //curl_setopt($ch, CURLOPT_PUT, 1);
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
        }
        if ($method === 'POST') {
            curl_setopt($ch, CURLOPT_POST, 1);
        }
        $body = curl_exec($ch);
        $statusCode = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        // echo "HTTP status code: " . $statusCode . PHP_EOL;
        // echo "HTTP response body length: " . strlen((string) $body) . PHP_EOL;
        if ($statusCode >= 400 || $statusCode < 200 || $body === false) {
            $curlError  = curl_error($ch);
            curl_close($ch);
            throw new \RuntimeException("HTTP request failed with status code {$statusCode}: {$curlError}");
        }
        curl_close($ch);

        return $body;
    }
}

class RetrieveTokenPayload
{
    public function __construct(
        private readonly string $apiKey,
        private readonly string $domain,
        private readonly string $legalName,
        private readonly string $vatId
    ) {}

    public function toArray(): array
    {
        return [
            'domain'     => $this->domain,
            'legalname'  => $this->legalName,
            'vatId'      => $this->vatId,
        ];
    }

    public function getApiKey(): string
    {
        return $this->apiKey;
    }
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

class People4InvoiceCli
{
    private string $jwt;

    public function __construct() {}

    // retrieve JWT token from Authentication API
    public function getToken(): string
    {
        $httpCli   = new People4HttpClient(disableSslVerification: true); // only for testing, not recommended for production
        $payload = new RetrieveTokenPayload(
            apiKey: "1c8f86e857cb65b20b26c481b0e2d2d1bf0b93a93aa7ccc35704c37ee63c597b",
            domain: "acme-corporation.eu",
            legalName: "ACME Corp",
            vatId: "NL8200.98.395.B.01",
        );
        $jwt = $httpCli->retrieveToken(
            url: 'https://app.people4.eu/auth-api/token',
            payload: $payload,
        );
        $this->jwt = $jwt;

        return $jwt;
    }

    // send invoice payload to Invoice API POST method
    public function createUblInvoice($invoiceJsonRaw): array
    {
        $httpCli   = new People4HttpClient(disableSslVerification: true); // only for testing, not recommended for production
        $invoiceCreated = $httpCli->sendInvoice(
            url: 'https://app.people4.eu/invoice-api/invoice/v1',
            payload: $invoiceJsonRaw,
            jwt: $this->jwt,
        );
        return $invoiceCreated;
    }
}

// example
try {
    $easyComplianceCli = new People4InvoiceCli();
    echo "Trying to retrieve JWT token from Authentication API" . PHP_EOL;
    $jwt = $easyComplianceCli->getToken();
    // echo "OK: JWT token retrieved: '" . $jwt . "'" . PHP_EOL;

    echo "Read json invoice minimal test file and send to People4 Invoice API" . PHP_EOL;
    $invoiceFile = __DIR__ . '/../inputs/invoice-v1-minimal-test.json';
    $invoiceJsonRaw = file_get_contents($invoiceFile);
    echo "Trying to send invoice payload to Invoice API to generate UBL and PEPPOL invoice" . PHP_EOL;
    $invoiceCreated = $easyComplianceCli->createUblInvoice($invoiceJsonRaw);
    // echo "DEBUG: Invoice created response: " . json_encode($invoiceCreated, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE) . PHP_EOL;
    $invoiceCreated['people4Id'] ?? throw new \RuntimeException("People4 ID not present in response.");
    $invoiceCreated['ubl'] ?? throw new \RuntimeException("UBL not present in response.");
    $invoiceCreated['peppol'] ?? throw new \RuntimeException("PEPPOL not present in response.");
    if (!empty($invoiceCreated['messages']) && is_array($invoiceCreated['messages'])) {
        foreach ($invoiceCreated['messages'] as $message) {
            echo "Error Message: " . $message . PHP_EOL;
        }
        throw new \RuntimeException("Error on generate UBL or PEPPOL message: " . $message);
    }
    if (isset($invoiceCreated['people4Id'])) {
        echo "OK: Invoice created with People4 ID: " . $invoiceCreated['people4Id'] . PHP_EOL;
    }
    if (isset($invoiceCreated['ubl'])) {
        echo "OK: UBL invoice created with length: " . strlen($invoiceCreated['ubl']) . PHP_EOL;
    }
    if (isset($invoiceCreated['peppol'])) {
        echo "OK: PEPPOL invoice created with length: " . strlen($invoiceCreated['peppol']) . PHP_EOL;
    }
    file_put_contents(__DIR__ . '/../outputs/response-invoice-v1-minimal-created.json', json_encode($invoiceCreated, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE));
    file_put_contents(__DIR__ . '/../outputs/response-invoice-v1-minimal-test-ubl.xml', $invoiceCreated['ubl']);
    file_put_contents(__DIR__ . '/../outputs/response-invoice-v1-minimal-test-peppol.xml', $invoiceCreated['peppol']);
    return 0;
} catch (\Throwable $e) {
    fwrite(STDERR, 'Fatal: ' . $e->getMessage() . PHP_EOL);
    echo "Error: " . $e->getMessage() . PHP_EOL;
    return 1;
}
